### PR TITLE
Use std::memcpy in USMObjectMem instead of placment new

### PR DESCRIPTION
**Pull request by masterleinad**
Alternative to #4266 should also fix #4254.

### CODE REVIEW

**Comment by masterleinad** [comment]
With this pull request, I am seeing (`devcloud`)
```
(Gen9, device)
N 1 M 10000 R 0 V 1 parallel_for: batched: 93.139321 single: 103.733955 fence: 4.247800e-05
N 1 M 10000 R 0 V 16 parallel_for: batched: 93.092840 single: 102.688207 fence: 4.331400e-05
N 1 M 10000 R 0 V 200 parallel_for: batched: 95.166861 single: 107.370191 fence: 4.330000e-05
N 1 M 10000 R 0 V 3000 parallel_for: batched: 119.198433 single: 116.933276 fence: 4.032900e-05
N 1 M 10000 R 0 V 30000 parallel_for: batched: 148.972830 single: 131.476627 fence: 4.012000e-05
```
```
(Gen9, shared)
N 1 M 10000 R 0 V 1 parallel_for: batched: 61.880885 single: 53.643341 fence: 4.475500e-05
N 1 M 10000 R 0 V 16 parallel_for: batched: 62.863985 single: 53.948255 fence: 4.499200e-05
N 1 M 10000 R 0 V 200 parallel_for: batched: 61.863525 single: 54.010082 fence: 4.363200e-05
N 1 M 10000 R 0 V 3000 parallel_for: batched: 62.404226 single: 53.300944 fence: 4.319000e-05
N 1 M 10000 R 0 V 30000 parallel_for: batched: 68.441915 single: 70.156949 fence: 3.682700e-05
```
whereas #4266 gives
```
(Gen9, device)
N 1 M 10000 R 0 V 1 parallel_for: batched: 92.829861 single: 106.146137 fence: 4.387400e-05
N 1 M 10000 R 0 V 16 parallel_for: batched: 92.583246 single: 106.147092 fence: 4.349400e-05
N 1 M 10000 R 0 V 200 parallel_for: batched: 97.959456 single: 108.308171 fence: 4.397200e-05
N 1 M 10000 R 0 V 3000 parallel_for: batched: 117.019304 single: 114.707066 fence: 4.029900e-05
N 1 M 10000 R 0 V 30000 parallel_for: batched: 155.048681 single: 135.398256 fence: 4.239900e-05
```
```
(Gen9, shared)
N 1 M 10000 R 0 V 1 parallel_for: batched: 63.430345 single: 55.500043 fence: 4.485700e-05
N 1 M 10000 R 0 V 16 parallel_for: batched: 63.482224 single: 55.465182 fence: 4.471100e-05
N 1 M 10000 R 0 V 200 parallel_for: batched: 63.029196 single: 55.286554 fence: 4.421800e-05
N 1 M 10000 R 0 V 3000 parallel_for: batched: 63.522038 single: 55.382384 fence: 4.355700e-05
N 1 M 10000 R 0 V 30000 parallel_for: batched: 82.135941 single: 309.468871 fence: 3.815100e-05
```

**Comment by masterleinad** [comment]
Given the performance results here and the simplifications over #4266, I am in favor of rather fixing #4254 by calling std::memcpy instead of using placement new in `copy_from` if the storage location is in shared memory.
This pull request now also removes member functions that were not used.

**Review by crtrott — APPROVED** [review]

**Review by nliber — APPROVED** [review]

**Comment by dalg24** [comment]
This merged changes from #4254 leaving a number of review comments unresolved...

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp [hunk 1]
```diff
@@ -136,97 +135,31 @@ class SYCLInternal {
...
-
-   private:
-    // Returns a reference to t (helpful when debugging)
-    template <typename T>
-    T& memcpy_to(T& t) {
-      assert(sizeof(T) == m_size);
-
-      sycl::event memcopied = m_q->memcpy(std::addressof(t), m_data, sizeof(T));
-      fence(memcopied,
+      fence();
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:143]: No label?

Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:143]: I needed to introduce it in #4266 and liked having it even if it's only used in one place. I would not have fought for it if someone didn't want to have it.

Comment (dalg24) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:143]: Well, it is a bit late to request a change

**Pull request merged**